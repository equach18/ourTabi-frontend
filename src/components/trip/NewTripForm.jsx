import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import Alert from "../common/Alert";

function NewTripForm({ addTrip }) {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    title: "",
    destination: "",
    radius: 10,
    startDate: "",
    endDate: "",
    isPrivate: true, // Default to private
  });
  const [formErrors, setFormErrors] = useState([]);
  const [isSubmitting, setIsSubmitting] = useState(false);

  /** Handle form field change */
  function handleChange(evt) {
    const { name, value, type, checked } = evt.target;
    setFormData((data) => ({
      ...data,
      [name]: type === "checkbox" ? checked : value,
    }));
  }

  /** Handle form submit */
  async function handleSubmit(evt) {
    evt.preventDefault();
    setFormErrors([]);
    setIsSubmitting(true);

    if (
      formData.startDate &&
      new Date(formData.startDate) >= new Date(formData.endDate)
    ) {
      setFormErrors(["End date must be after start date."]);
      setIsSubmitting(false);
      return;
    }

    try {
      const result = await addTrip(formData);
      if (result.success) {
        navigate(`/trips/${result.trip.id}`);
      } else {
        setFormErrors([
          result.error?.message ||
            String(result.error || "Failed to create trip."),
        ]);
      }
    } catch (err) {
      setFormErrors([err.message || String(err)]);
    } finally {
      setIsSubmitting(false);
    }
  }

  return (
    <div className="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10">
      <div className="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg">
        <h1 className="text-3xl font-bold text-gray-600 mb-4">
          Create A New Trip
        </h1>

        <form onSubmit={handleSubmit} className="space-y-3">
          {/* title */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="title">
              Title
            </label>
            <input
              type="text"
              name="title"
              id="title"
              value={formData.title}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required
            />
          </div>

          {/* destination */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="destination">
              Destination
            </label>
            <input
              type="text"
              name="destination"
              id="destination"
              value={formData.destination}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required
            />
          </div>

          {/* radius */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="radius">
              Radius (miles/km)
            </label>
            <input
              type="number"
              name="radius"
              id="radius"
              value={formData.radius}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              min="1"
              required
            />
          </div>

          {/* start date */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="startDate">
              Start Date
            </label>
            <input
              type="date"
              name="startDate"
              id="startDate"
              value={formData.startDate}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            />
          </div>

          {/* end date */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="endDate">
              End Date
            </label>
            <input
              type="date"
              name="endDate"
              id="endDate"
              value={formData.endDate}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required={!!formData.startDate}
              disabled={!formData.startDate}
            />
          </div>

          {/* privacy */}
          <div className="flex items-center space-x-2">
            <input
              type="checkbox"
              name="isPrivate"
              id="privacy"
              checked={formData.isPrivate}
              onChange={handleChange}
              className="w-4 h-4 text-emerald-600 border-gray-300 rounded focus:ring-emerald-500"
            />
            <label className="text-zinc-700" htmlFor="privacy">Private Trip</label>
          </div>

          {/* error alerts */}
          <Alert type="error" messages={formErrors} />

          <button
            type="submit"
            className={`w-full bg-blue-500 text-white py-2 rounded-lg hover:bg-blue-600 transition ${
              isSubmitting && "opacity-50 cursor-not-allowed"
            }`}
            disabled={isSubmitting}
          >
            {isSubmitting ? "Creating..." : "Create Trip"}
          </button>
        </form>
      </div>
    </div>
  );
}

export default NewTripForm;
