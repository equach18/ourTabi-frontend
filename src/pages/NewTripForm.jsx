import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import Alert from "../common/Alert";

function NewTripForm({ addTrip }) {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    title: "",
    destination: "",
    radius: 10,
    startDate: "",
    endDate: "",
    isPrivate: true, // Default to private
  });
  const [formErrors, setFormErrors] = useState([]);
  const [isSubmitting, setIsSubmitting] = useState(false);

  /** Handle form field change */
  function handleChange(evt) {
    const { name, value, type, checked } = evt.target;
    setFormData((data) => ({
      ...data,
      [name]: type === "checkbox" ? checked : value,
    }));
  }

  /** Handle form submit */
  async function handleSubmit(evt) {
    evt.preventDefault();
    setFormErrors([]);
    setIsSubmitting(true);

    if (
      formData.startDate &&
      new Date(formData.startDate) >= new Date(formData.endDate)
    ) {
      setFormErrors(["End date must be after start date."]);
      setIsSubmitting(false);
      return;
    }

    try {
      const result = await addTrip(formData); 
      if (result.success) {
        navigate(`/trips/${result.trip.id}`);
      } else{
        setFormErrors([result.error])
      }
    } catch (err) {
      setFormErrors([err]);
    } finally {
      setIsSubmitting(false);
    }
  }

  return (
    <div className="max-w-lg mx-auto mt-10 p-6 bg-white shadow-lg rounded-lg">
      <h1 className="text-2xl font-bold text-blue-500">Create a New Trip</h1>

      <form onSubmit={handleSubmit} className="mt-4">
        {/* Title */}
        <div className="mb-4">
          <label className="block text-gray-700">Title</label>
          <input
            type="text"
            name="title"
            value={formData.title}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            required
          />
        </div>

        {/* Destination */}
        <div className="mb-4">
          <label className="block text-gray-700">Destination</label>
          <input
            type="text"
            name="destination"
            value={formData.destination}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            required
          />
        </div>

        {/* Radius */}
        <div className="mb-4">
          <label className="block text-gray-700">Radius (miles/km)</label>
          <input
            type="number"
            name="radius"
            value={formData.radius}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            min="1"
            required
          />
        </div>

        {/* Start Date (optional) */}
        <div className="mb-4">
          <label className="block text-gray-700">Start Date</label>
          <input
            type="date"
            name="startDate"
            value={formData.startDate}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
          />
        </div>

        {/* End Date (required only if startDate exists) */}
        <div className="mb-4">
          <label className="block text-gray-700">End Date</label>
          <input
            type="date"
            name="endDate"
            value={formData.endDate}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            required={!!formData.startDate} // Require only if startDate exists
            disabled={!formData.startDate} // Disable unless startDate is filled
          />
        </div>

        {/* Private Trip Checkbox */}
        <div className="mb-4">
          <label className="inline-flex items-center">
            <input
              type="checkbox"
              name="isPrivate"
              checked={formData.isPrivate}
              onChange={handleChange}
              className="mr-2"
            />
            Private Trip
          </label>
        </div>

        {/* Error Messages */}
        {formErrors.length > 0 && <Alert type="error" messages={formErrors} />}

        {/* Submit Button */}
        <button
          type="submit"
          className="bg-blue-500 text-white px-6 py-3 rounded-lg hover:bg-blue-600 w-full"
          disabled={isSubmitting}
        >
          {isSubmitting ? "Creating..." : "Create Trip"}
        </button>
      </form>
    </div>
  );
}

export default NewTripForm;
