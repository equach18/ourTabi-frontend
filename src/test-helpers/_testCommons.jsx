import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
