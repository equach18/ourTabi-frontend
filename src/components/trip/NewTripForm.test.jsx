import { fireEvent, screen, waitFor } from "@testing-library/react";
import { Routes, Route } from "react-router-dom";
import NewTripForm from "./NewTripForm";
import { renderWithAllProviders } from "../../test-helpers/_testCommons";
const mockNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

describe("NewTripForm", () => {
  const mockAddTrip = vi.fn();

  function renderForm() {
    return renderWithAllProviders(
      <Routes>
        <Route
          path="/trips/new"
          element={<NewTripForm addTrip={mockAddTrip} />}
        />
      </Routes>,
      { route: "/trips/new" }
    );
  }

  beforeEach(() => {
    mockAddTrip.mockReset();
    mockNavigate.mockReset();
  });

  it("renders correct form fields", () => {
    renderForm();
    expect(screen.getByLabelText(/title/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/destination/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/radius/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/start date/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/end date/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/private trip/i)).toBeInTheDocument();
  });
  
  it("matches snapshot", () => {
    const { asFragment } = renderForm();
    expect(asFragment()).toMatchSnapshot();
  });

  it("disables end date if start date is not selected", () => {
    renderForm();
    const endDate = screen.getByLabelText(/end date/i);
    expect(endDate).toBeDisabled();
  });

  it("calls addTrip navigates successful submission", async () => {
    mockAddTrip.mockResolvedValue({
      success: true,
      trip: { id: 99 },
    });

    renderForm();

    fireEvent.change(screen.getByLabelText(/title/i), {
      target: { value: "test title" },
    });

    fireEvent.change(screen.getByLabelText(/destination/i), {
      target: { value: "test destination" },
    });

    fireEvent.change(screen.getByLabelText(/radius/i), {
      target: { value: "14" },
    });

    fireEvent.click(screen.getByRole("button", { name: /create trip/i }));

    await waitFor(() => {
      expect(mockAddTrip).toHaveBeenCalled();
      expect(mockNavigate).toHaveBeenCalledWith("/trips/99");
    });
  });

  it("displays error if addTrip fails", async () => {
    mockAddTrip.mockResolvedValue({
      success: false,
      error: { message: "Mock API error" },
    });

    renderForm();

    fireEvent.change(screen.getByLabelText(/title/i), {
      target: { value: "Fail Trip" },
    });

    fireEvent.change(screen.getByLabelText(/destination/i), {
      target: { value: "Nowhere" },
    });

    fireEvent.click(screen.getByRole("button", { name: /create trip/i }));

    expect(await screen.findByText(/mock api error/i)).toBeInTheDocument();
  });
});
