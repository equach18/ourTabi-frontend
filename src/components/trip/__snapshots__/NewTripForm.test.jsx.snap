// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewTripForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10"
  >
    <div
      class="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg"
    >
      <h1
        class="text-3xl font-bold text-gray-600 mb-4"
      >
        Create A New Trip
      </h1>
      <form
        class="space-y-3"
      >
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="title"
          >
            Title
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="title"
            name="title"
            required=""
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="destination"
          >
            Destination
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="destination"
            name="destination"
            required=""
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="radius"
          >
            Radius (miles/km)
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="radius"
            min="1"
            name="radius"
            required=""
            type="number"
            value="10"
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="startDate"
          >
            Start Date
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="startDate"
            name="startDate"
            type="date"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="endDate"
          >
            End Date
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            disabled=""
            id="endDate"
            name="endDate"
            type="date"
            value=""
          />
        </div>
        <div
          class="flex items-center space-x-2"
        >
          <input
            checked=""
            class="w-4 h-4 text-emerald-600 border-gray-300 rounded focus:ring-emerald-500"
            id="privacy"
            name="isPrivate"
            type="checkbox"
          />
          <label
            class="text-zinc-700"
            for="privacy"
          >
            Private Trip
          </label>
        </div>
        <button
          class="w-full bg-blue-500 text-white py-2 rounded-lg hover:bg-blue-600 transition false"
          type="submit"
        >
          Create Trip
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
